### PR TITLE
feat(cross-chain-oracle): Add Polygon adapters

### PR DESCRIPTION
**Pull request by nicholaspai**
**Motivation**

These adapters re use contracts from `external/polygon` that currently enable the cross chain contracts in `contracts/polygon` that only work for Polygon.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [X]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

### CODE REVIEW

File: packages/core/contracts/cross-chain-oracle/chain-adapters/Optimism_ChildMessenger.sol [hunk 1]
```diff
@@ -53,7 +54,11 @@ contract Optimism_ChildMessenger is OVM_CrossDomainEnabled, ChildMessengerInterf
      * @dev The caller of this function must be the parent messenger, over the canonical bridge.
      * @param newParentMessenger address of the new parent messenger, deployed on L1.
      */
-    function setParentMessenger(address newParentMessenger) public onlyFromCrossDomainAccount(parentMessenger) {
+    function setParentMessenger(address newParentMessenger)
+        public
+        onlyFromCrossDomainAccount(parentMessenger)
+        nonReentrant()
```
Comment (chrismaree) [inline packages/core/contracts/cross-chain-oracle/chain-adapters/Optimism_ChildMessenger.sol:60]: I think these `nonReentrant` modifies are redundant (and potentially dangerous) for us to user here. these methods are access controlled via the `onlyFromCrossDomainAccount` modifier so why do we need this entry guard?

File: packages/core/contracts/cross-chain-oracle/chain-adapters/Polygon_ChildMessenger.sol [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import "../../common/implementation/Lockable.sol";
+
+/**
+ * @notice Sends cross chain messages from Polygon to Ethereum network.
+ * @dev This contract extends the `FxBaseChildTunnel` contract and therefore is 1-to-1 mapped with the
+ * `FxBaseRootTunnel` extended by the `Polygon_ParentMessenger` contract deployed on Polygon. This mapping ensures that
+ * the internal `_processMessageFromRoot` function is only callable indirectly by the `Polygon_ParentMessenger`.
+ */
+contract Polygon_ChildMessenger is FxBaseChildTunnel, ChildMessengerInterface, Lockable {
+    // The only child network contract that can send messages over the bridge via the messenger is the oracle spoke.
```
Comment (chrismaree) [inline packages/core/contracts/cross-chain-oracle/chain-adapters/Polygon_ChildMessenger.sol:16]:
```suggestion
    // The only child network contract that can send messages over the bridge via the messenger is the Oracle Spoke.
```

Comment (chrismaree) [reply packages/core/contracts/cross-chain-oracle/chain-adapters/Polygon_ChildMessenger.sol:16]: same nit throughout

File: packages/core/contracts/cross-chain-oracle/chain-adapters/Polygon_ChildMessenger.sol [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    /**
+     * @notice Sends a message to the OracleSpoke via the parent messenger and the canonical message bridge.
+     * @dev The caller must be the OracleSpoke on child network. No other contract is permissioned to call this
+     * function.
+     * @dev The L1 target, the parent messenger, must implement processMessageFromChild to consume the message.
+     * @param data data message sent to the L1 messenger. Should be an encoded function call or packed data.
+     */
+    function sendMessageToParent(bytes memory data) public override nonReentrant() {
+        require(msg.sender == oracleSpoke, "Only callable by oracleSpoke");
+        _sendMessageToRoot(abi.encode(data, oracleHub));
```
Comment (chrismaree) [inline packages/core/contracts/cross-chain-oracle/chain-adapters/Polygon_ChildMessenger.sol:65]: hhmmm. is this correct? it might be an artefact of polygon that I'm not fully understanding. why are you not calling the call data something akin to the optimism one like this:
```
bytes memory dataSentToParent = abi.encodeWithSignature("processMessageFromCrossChainChild(bytes)", data);
```
where we encode the function that is going to be called on the `ParentMessenger`

File: packages/core/contracts/cross-chain-oracle/chain-adapters/Polygon_ChildMessenger.sol [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+     * and Polygon. This will revert if the Root chain sender is not the `fxRootTunnel` contract.
+     * @dev This call will revert if `setFxRoot` has not been called. Fx Child should be set to Polygon_ParentMessenger.
+     * @param sender The sender of `data` from the Root chain.
+     * @param data ABI encoded params with which to call function on OracleHub or GovernorHub.
+     */
+    function _processMessageFromRoot(
+        uint256, /* stateId */
+        address sender,
+        bytes memory data
+    ) internal override validateSender(sender) nonReentrant() {
```
Comment (chrismaree) [inline packages/core/contracts/cross-chain-oracle/chain-adapters/Polygon_ChildMessenger.sol:81]: where is this `validateSender` coming from? how does that work? a comment might be needed as I cant find this in the PR anywhere

File: packages/core/contracts/cross-chain-oracle/chain-adapters/test/Polygon_ChildMessengerMock.sol [hunk 5]
```diff
@@ -0,0 +1,16 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.0;
+
+import "../Polygon_ChildMessenger.sol";
+
+contract Polygon_ChildMessengerMock is Polygon_ChildMessenger {
```
Comment (chrismaree) [inline packages/core/contracts/cross-chain-oracle/chain-adapters/test/Polygon_ChildMessengerMock.sol:6]: kinda lame that you need extra mocks for polygon.

File: packages/core/contracts/cross-chain-oracle/interfaces/ChildMessengerInterface.sol [hunk 6]
```diff
@@ -4,8 +4,4 @@ pragma solidity ^0.8.0;
 interface ChildMessengerInterface {
     // Should send cross-chain message to Parent messenger contract or revert.
     function sendMessageToParent(bytes memory data) external;
-
-    // Should be called by parent messenger over the canonical bridge. Includes a target as the L2 contract that should
-    // have data passed to it.
-    function processMessageFromCrossChainParent(bytes memory data, address target) external;
```
Comment (chrismaree) [inline packages/core/contracts/cross-chain-oracle/interfaces/ChildMessengerInterface.sol:10]: why did you remove this?

File: packages/core/contracts/cross-chain-oracle/chain-adapters/Polygon_ParentMessenger.sol [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+    /**
+     * @notice Process a received message from the child messenger via the canonical message bridge.
+     * @dev This internal method will be called inside `FxBaseRootTunnel.receiveMessage(bytes memory inputData)`.
+     * The `inputData` is a proof of transaction that is derived from the transaction hash of the transaction on the
+     * child chain that originated the cross-chain price request via _sendMessageToRoot.
+     * @dev This call will revert if `setFxChild` has not been called. Fx Child should be set to Polygon_ChildMessenger.
+     * @param data ABI encoded params with which to call function on OracleHub or GovernorHub.
+     */
+    function _processMessageFromChild(bytes memory data) internal override nonReentrant() {
```
Comment (chrismaree) [inline packages/core/contracts/cross-chain-oracle/chain-adapters/Polygon_ParentMessenger.sol:57]: why is this internal? how does the polygon Fxtunnel work with this to know to call this function?

File: packages/core/contracts/cross-chain-oracle/chain-adapters/Optimism_ChildMessenger.sol [hunk 8]
```diff
@@ -91,10 +100,10 @@ contract Optimism_ChildMessenger is OVM_CrossDomainEnabled, ChildMessengerInterf
      */
     function processMessageFromCrossChainParent(bytes memory data, address target)
         public
-        override
```
Comment (chrismaree) [inline packages/core/contracts/cross-chain-oracle/chain-adapters/Optimism_ChildMessenger.sol:94]: why is this no longer override? do we not want this to be the standard interface for inbound messages from the parent -> child?

Comment (chrismaree) [reply packages/core/contracts/cross-chain-oracle/chain-adapters/Optimism_ChildMessenger.sol:94]: same question on the parent in the following file.

File: packages/core/networks/5.json [hunk 9]
```diff
@@ -22,5 +22,21 @@
   {
     "contractName": "MockOracleAncillary",
     "address": "0x20570e9e27920ac5e2601e0bef7244deff7f0b28"
+  },
+  {
+    "contractName": "ExpandedERC20",
```
Comment (chrismaree) [inline packages/core/networks/5.json:27]: why do we need this?

File: packages/core/contracts/cross-chain-oracle/chain-adapters/Optimism_ChildMessenger.sol (hunk 1 above)
Comment (nicholaspai) [reply packages/core/contracts/cross-chain-oracle/chain-adapters/Optimism_ChildMessenger.sol:60]: I don't see that they're dangerous, perhaps they're overkill, but why would reentrancy ever happen?

File: packages/core/contracts/cross-chain-oracle/chain-adapters/Polygon_ChildMessenger.sol (hunk 3 above)
Comment (nicholaspai) [reply packages/core/contracts/cross-chain-oracle/chain-adapters/Polygon_ChildMessenger.sol:65]: Yeah this is polygon specific. When sending a message from Polygon to Ethereum, the Polygon contract must emit `MessageSent`. Then, the user can manually relay the message when the block containing the transaction with the `MessageSent` event is checkpointed to Mainnet.

To manually relay the message, the user must call `receiveMessage` on the `Polygon_ParentMessenger` or whatever the `FxBaseRootTunnel` that is connected to this `FxBaseChildTunnel`. `receiveMessage` will then call the internal `_processMessageFromChild` method. This is why we don't need to use the `abi.encodeWithSignature` because we know already which internal method on the `ParentMessenger` will be called.

File: packages/core/contracts/cross-chain-oracle/chain-adapters/Polygon_ChildMessenger.sol (hunk 4 above)
Comment (nicholaspai) [reply packages/core/contracts/cross-chain-oracle/chain-adapters/Polygon_ChildMessenger.sol:81]: Its in `external/polygon/FxBaseChildTunnel`.  I edited the comments.

File: packages/core/contracts/cross-chain-oracle/chain-adapters/test/Polygon_ChildMessengerMock.sol (hunk 5 above)
Comment (nicholaspai) [reply packages/core/contracts/cross-chain-oracle/chain-adapters/test/Polygon_ChildMessengerMock.sol:6]: That's only because there are internal methods to test

File: packages/core/contracts/cross-chain-oracle/interfaces/ChildMessengerInterface.sol (hunk 6 above)
Comment (nicholaspai) [reply packages/core/contracts/cross-chain-oracle/interfaces/ChildMessengerInterface.sol:10]: Because we do not need it in `Polygon` messengers, and I see no harm in adding it to the Optimism adapters

File: packages/core/contracts/cross-chain-oracle/chain-adapters/Polygon_ParentMessenger.sol (hunk 7 above)
Comment (nicholaspai) [reply packages/core/contracts/cross-chain-oracle/chain-adapters/Polygon_ParentMessenger.sol:57]: This is how Polygon works. I tried to explain in comments can you let me know what could be changed about this language>

File: packages/core/contracts/cross-chain-oracle/chain-adapters/Optimism_ChildMessenger.sol (hunk 8 above)
Comment (nicholaspai) [reply packages/core/contracts/cross-chain-oracle/chain-adapters/Optimism_ChildMessenger.sol:94]: I don't think this is standard for all networks. Only Optimism so far, and Arbitrum will work the same. But Polygon works differently so I don't think it should go in the interface

File: packages/core/networks/5.json (hunk 9 above)
Comment (nicholaspai) [reply packages/core/networks/5.json:27]: We don't I'll remove

**Review by chrismaree — APPROVED** [review]
